### DISCUSSION TITLE
Improving `viewStore.binding(get:send:)` ergonomics

### DISCUSSION THREAD

**Discussion by Jeehut** [pointfreeco/swift-composable-architecture#1054] [Ideas]
# The Problem
I regularly run into issues with the `viewStore.binding(get:send:)` method. The two issues I have all the time:

1. I forget what the parameter types of the `get` and `send` are and what the expected return type of `send` is.
2. I get really long build times resulting in an `The compiler is unable to type-check this expression in reasonable time` error.

For example, see this code with which I struggled quite a bit to figure out what the problem was:

```swift
viewStore.binding(
  get: { $0.unlockedTranslations.contains(translation.id) },
  send: { AddTranslationAction.translationLockChanged(translationId: translation.id, locked: $0) }
)
```

Even adding the types explicitly didn't help the compiler to figure out the issue, I still received the same error:
```swift
viewStore.binding(
  get: { (state: AddTranslationState) -> Bool in state.unlockedTranslations.contains(translation.id) },
  send: { (newValue: Bool) -> AddTranslationAction in AddTranslationAction.translationLockChanged(translationId: translation.id, locked: newValue) }
)
```

# My Solution
I created myself this extension, which fixes 1. by providing more speaking names for the parameters as to what is passed and expected and fixes 2. with an explicit `valueType` parameter to help the compiler figure out the generic type here:

```swift
import ComposableArchitecture
import SwiftUI

extension ViewStore {
  public func bind<Value>(
    valueFromState: @escaping (State) -> Value,
    actionFromValue: @escaping (Value) -> Action,
    valueType: Value.Type
  ) -> Binding<Value> {
    self.binding(get: valueFromState, send: actionFromValue)
  }
}
```

Usage with it looked like this:

```swift
viewStore.bind(
  valueFromState: { $0.unlockedTranslations.contains(translation.id) }, // <-- Cannot convert value of type 'UUID' to expected argument type 'AddTranslationState.Translation'
  sendActionWithValue: { AddTranslationAction.translationLockChanged(translationId: translation.id, locked: $0) },
  valueType: Bool.self
)
```

Immediately this gave me an error that made clear what the issue was as Swift could now figure out the types easily: "Cannot convert value of type 'UUID' to expected argument type 'AddTranslationState.Translation'". By replacing `$0.unlockedTranslations` with `$0.lockedTranslationIds` I could fix this build error and now the compiler is "able to type-check this expression in reasonable time" – in less than a second even!

I'm sharing this to both start a discussion if the `binding(get:send:)` method should maybe get an overload in the library and also to help people running into the same problem with the above extension. The Swift compiler error doesn't really help.

**Comment by stephencelis** [comment]
@Jeehut The `binding` method is overloaded a few times, which could be the source of the problem when it comes to error messaging. While I'm sure we've encountered a similar issue in the past, we don't have a repro handy. If you have one to share we could potentially experiment with adding `@_disfavoredOverload` to some of the overloads to get better error messaging.

**Reply by Jeehut** [reply]
I just tried extracting the relevant code out for reproduction, but it's hard because as soon as I put that `Picker` where I use the binding for the `selection` parameter in a simpler context, the compiler actually seems to be able to figure things out. It really seems this only happens if the compiler is already getting close to that `unable to type-check` limit, in which case making things explicit with the `valueType` I had provided seems to help the compiler. To test out your theory, I also created this function alias:

```swift
extension ViewStore {
  public func customBinding<LocalState>(
    get: @escaping (State) -> LocalState,
    send localStateToViewAction: @escaping (LocalState) -> Action
  ) -> Binding<LocalState> {
    self.binding(get: get, send: localStateToViewAction)
  }
}

```
Calling this `customBinding` function then worked as well, so you might be right with the overload.

**Reply by Jeehut** [reply]
As I already mentioned in my original post, I personally find `get` and `send` to be not very clear names. For example, for the overload function `send: Action` I would find the name `action: Action` more clear (you pass an action directly), whereas I would use `actionFromValue: (LocalState) -> Action` for the closure variant (I would also rename `LocalState` to `Value` because in most cases we're just passing a value, not a `State` object). This way, it should also be easier for Swift to figure out which method you mean because with the paramater label change, also the function name changes and it's technically no overloading anymore AFAIK. Just something to think about.

**Comment by Jeehut** [comment]
After @stephencelis's [feedback](https://github.com/pointfreeco/swift-composable-architecture/discussions/1054#discussioncomment-2523732) I changed my extension to this:

```swift
extension ViewStore {
  public func binding<Value>(
    valueFromState: @escaping (State) -> Value,
    actionFromValue: @escaping (Value) -> Action
  ) -> Binding<Value> {
    self.binding(get: valueFromState, send: actionFromValue)
  }

  public func binding<Value>(
    valueFromState: @escaping (State) -> Value,
    action: Action
  ) -> Binding<Value> {
    self.binding(get: valueFromState, send: action)
  }
}

```

Just in case somebody wanted to copy & paste.
